relay-replays/src/recording.rs

```diff
@@ -68,31 +68,72 @@ impl From<serde_json::Error> for ParseRecordingError {
     }
 }
 
-static STRING_STATE: Lazy<ProcessingState> = Lazy::new(|| {
-    ProcessingState::root().enter_static(
-        "",
-        Some(Cow::Owned(FieldAttrs::new().pii(Pii::True))),
-        Some(ValueType::String),
-    )
+/// Paths to fields on which datascrubbing rules should be applied.
+///
+/// This is equivalent to marking a field as `pii = true` in an `Annotated` schema.
+static PII_FIELDS: Lazy<[Vec<&str>; 2]> = Lazy::new(|| {
+    [
+        vec!["data", "payload", "description"],
+        vec!["data", "payload", "data"],
+    ]
 });
 
+/// Returns `True` if the given path should be treated as `pii = true`.
+fn scrub_at_path(path: &Vec<String>) -> bool {
+    PII_FIELDS
+        .iter()
+        .any(|p| p.iter().zip(path).all(|(k1, k2)| k1 == k2))
+}
+
+/// Static field attributes used for every field.
+const FIELD_ATTRS_PII_TRUE: FieldAttrs = FieldAttrs::new().pii(Pii::True);
+const FIELD_ATTRS_PII_FALSE: FieldAttrs = FieldAttrs::new().pii(Pii::False);
+
 /// The [`Transform`] implementation for data scrubbing.
 ///
 /// This is used by [`EventStreamVisitor`] and [`ScrubbedValue`] to scrub recording events.
 struct ScrubberTransform<'a> {
+    /// PII processors that are applied one by one on each value.
     processor1: Option<PiiProcessor<'a>>,
     processor2: Option<PiiProcessor<'a>>,
+    /// The state encoding the current path, which is fed by `push_path` and `pop_path`.
+    state: ProcessingState<'a>,
+    /// The current path. This is redundant with `state`, which also contains the full path,
+    /// but easier to match on.
+    path: Vec<String>,
 }
 
-impl Transform for &'_ mut ScrubberTransform<'_> {
+impl<'de> Transform<'de> for &'_ mut ScrubberTransform<'_> {
+    fn push_path(&mut self, key: &'de str) {
+        self.path.push(key.to_owned());
+        let field_attrs = if scrub_at_path(&self.path) {
+            &FIELD_ATTRS_PII_TRUE
+        } else {
+            &FIELD_ATTRS_PII_FALSE
+        };
+
+        self.state = std::mem::take(&mut self.state).enter_owned(
+            key.to_owned(),
+            Some(Cow::Borrowed(field_attrs)),
```

Comment:
Q: this sets the `field_attrs`, but there may be additional fields that are being overwritten and are now lost; is this correct? In other words, should we overwrite or merge the attributes? It's also possible this question does not make sense at all.

Reply:
Previously, we just used the static `STRING_STATE` here, which hard-coded its field attrs to `FieldAttrs::new().pii(Pii::True)` ([link](https://github.com/getsentry/relay/pull/2034/files/f67bef1b33b86a7db0acb204f3d66597e704d59b#diff-37853ccc35bc20c05e8c5a3f04679a82d62c0c1efb959f8db3e42049c97b6b16L74)).

Since replays are schemaless, all we care about for the purpose of data scrubbing is toggling the pii field attr.